File: services/cloud/internal/contracts/card_definition_test.go
package contracts_test

import (
	"errors"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"

	"github.com/zzq/agent-card-container/services/cloud/internal/contracts"
)

func TestDecodeCardDefinitionFixtures(t *testing.T) {
	t.Parallel()

	tests := []struct {
		name       string
		fixture    string
		runtime    contracts.CardRuntime
		entrypoint string
		capability string
	}{
		{
			name:       "native",
			fixture:    "native-card.json",
			runtime:    contracts.CardRuntimeNative,
			entrypoint: "payload/native.json",
			capability: "notification.show",
		},
		{
			name:       "web",
			fixture:    "web-card.json",
			runtime:    contracts.CardRuntimeWeb,
			entrypoint: "payload/web/index.html",
			capability: "storage",
		},
	}

	for _, tt := range tests {
		tt := tt
		t.Run(tt.name, func(t *testing.T) {
			t.Parallel()

			definition, err := contracts.DecodeCardDefinition(
				strings.NewReader(string(readFixture(t, tt.fixture))),
			)
			if err != nil {
				t.Fatalf("DecodeCardDefinition() error = %v", err)
			}

			if definition.Runtime != tt.runtime {
				t.Fatalf("Runtime = %q, want %q", definition.Runtime, tt.runtime)
			}
			if definition.Entrypoint != tt.entrypoint {
				t.Fatalf("Entrypoint = %q, want %q", definition.Entrypoint, tt.entrypoint)
			}
			if definition.PreferredSize.Width <= definition.MinSize.Width {
				t.Fatalf("PreferredSize.Width = %v, want greater than %v", definition.PreferredSize.Width, definition.MinSize.Width)
			}
			if !definition.HasCapability(tt.capability) {
				t.Fatalf("HasCapability(%q) = false", tt.capability)
			}
		})
	}
}

func TestDecodeCardDefinitionRejectsUnknownRuntime(t *testing.T) {
	t.Parallel()

	input := strings.Replace(
		string(readFixture(t, "web-card.json")),
		"\"runtime\": \"web\"",
		"\"runtime\": \"script\"",
		1,
	)

	_, err := contracts.DecodeCardDefinition(strings.NewReader(input))

	if !errors.Is(err, contracts.ErrInvalidRuntime) {
		t.Fatalf("error = %v, want ErrInvalidRuntime", err)
	}
}

func readFixture(t *testing.T, name string) []byte {
	t.Helper()

	_, currentFile, _, ok := runtime.Caller(0)
	if !ok {
		t.Fatal("runtime.Caller() failed")
	}
	path := filepath.Join(
		filepath.Dir(currentFile),
		"..", "..", "..", "..",
		"contracts", "card", "fixtures", name,
	)
	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read fixture %q: %v", name, err)
	}
	return data
}

File: apps/desktop/test/contracts/card_definition_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:agent_card_desktop/src/contracts/card_definition.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  group('CardDefinition', () {
    for (final testCase
        in <
          ({
            String fixture,
            CardRuntime runtime,
            String entrypoint,
            String capability,
          })
        >[
          (
            fixture: 'native-card.json',
            runtime: CardRuntime.native,
            entrypoint: 'payload/native.json',
            capability: 'notification.show',
          ),
          (
            fixture: 'web-card.json',
            runtime: CardRuntime.web,
            entrypoint: 'payload/web/index.html',
            capability: 'storage',
          ),
        ]) {
      test('parses ${testCase.runtime.name} fixture', () {
        final definition = CardDefinition.fromJson(
          jsonDecode(_readFixture(testCase.fixture)) as Map<String, Object?>,
        );

        expect(definition.runtime, testCase.runtime);
        expect(definition.entrypoint, testCase.entrypoint);
        expect(
          definition.preferredSize.width,
          greaterThan(definition.minSize.width),
        );
        expect(definition.hasCapability(testCase.capability), isTrue);
      });
    }

    test('rejects an unknown runtime', () {
      final source = _readFixture(
        'web-card.json',
      ).replaceFirst('"runtime": "web"', '"runtime": "script"');

      expect(
        () =>
            CardDefinition.fromJson(jsonDecode(source) as Map<String, Object?>),
        throwsFormatException,
      );
    });
  });
}

String _readFixture(String name) {
  return File('../../contracts/card/fixtures/$name').readAsStringSync();
}
